// @capgo/capacitor-file-picker/dist/plugin.cjs.js
'use strict';

var core = require('@capacitor/core');

const CapgoFilePicker = core.registerPlugin('CapgoFilePicker', {
    web: () => Promise.resolve().then(function () { return web; }).then((m) => new m.CapgoFilePickerWeb()),
});

class CapgoFilePickerWeb extends core.WebPlugin {
    async pickFiles(options) {
        const files = await this.openFilePicker({
            types: options === null || options === void 0 ? void 0 : options.types,
            multiple: (options === null || options === void 0 ? void 0 : options.limit) !== 1,
        });
        const pickedFiles = [];
        for (const file of files) {
            const pickedFile = {
                name: file.name,
                mimeType: file.type || 'application/octet-stream',
                size: file.size,
                blob: file,
                modifiedAt: file.lastModified,
            };
            if (options === null || options === void 0 ? void 0 : options.readData) {
                pickedFile.data = await this.getDataFromFile(file);
            }
            if (file.type.startsWith('image/')) {
                const dimensions = await this.getImageDimensions(file);
                pickedFile.width = dimensions.width;
                pickedFile.height = dimensions.height;
            }
            pickedFiles.push(pickedFile);
        }
        return { files: pickedFiles };
    }
    async pickImages(options) {
        return this.pickFiles({
            types: ['image/*'],
            limit: options === null || options === void 0 ? void 0 : options.limit,
            readData: options === null || options === void 0 ? void 0 : options.readData,
        });
    }
    async pickVideos(options) {
        return this.pickFiles({
            types: ['video/*'],
            limit: options === null || options === void 0 ? void 0 : options.limit,
            readData: options === null || options === void 0 ? void 0 : options.readData,
        });
    }
    async pickMedia(options) {
        return this.pickFiles({
            types: ['image/*', 'video/*'],
            limit: options === null || options === void 0 ? void 0 : options.limit,
            readData: options === null || options === void 0 ? void 0 : options.readData,
        });
    }
    async pickDirectory() {
        throw this.unimplemented('pickDirectory is not supported on web');
    }
    async convertHeicToJpeg(_options) {
        throw this.unimplemented('convertHeicToJpeg is not supported on web');
    }
    async copyFile(_options) {
        throw this.unimplemented('copyFile is not supported on web');
    }
    async checkPermissions() {
        return {
            readExternalStorage: 'granted',
        };
    }
    async requestPermissions() {
        return {
            readExternalStorage: 'granted',
        };
    }
    async addListener(_eventName, _listenerFunc) {
        return {
            remove: async () => {
                // No-op on web
            },
        };
    }
    async removeAllListeners() {
        // No-op on web
    }
    async getPluginVersion() {
        return { version: 'web' };
    }
    openFilePicker(options) {
        return new Promise((resolve) => {
            var _a;
            const input = document.createElement('input');
            input.type = 'file';
            input.multiple = (_a = options.multiple) !== null && _a !== void 0 ? _a : true;
            if (options.types && options.types.length > 0) {
                input.accept = options.types.join(',');
            }
            input.style.display = 'none';
            document.body.appendChild(input);
            let resolved = false;
            const handleChange = () => {
                resolved = true;
                const files = Array.from(input.files || []);
                cleanup();
                resolve(files);
            };
            const handleCancel = () => {
                // Small delay to ensure change event fires first if files were selected
                setTimeout(() => {
                    if (!resolved) {
                        cleanup();
                        resolve([]);
                    }
                }, 300);
            };
            const cleanup = () => {
                input.removeEventListener('change', handleChange);
                window.removeEventListener('focus', handleCancel);
                document.body.removeChild(input);
            };
            input.addEventListener('change', handleChange);
            window.addEventListener('focus', handleCancel);
            input.click();
        });
    }
    getDataFromFile(file) {
        return new Promise((resolve, reject) => {
            const reader = new FileReader();
            reader.onload = () => {
                const result = reader.result;
                // Remove data URL prefix to get pure base64
                const base64 = result.split(',')[1] || '';
                resolve(base64);
            };
            reader.onerror = () => reject(new Error('Failed to read file'));
            reader.readAsDataURL(file);
        });
    }
    getImageDimensions(file) {
        return new Promise((resolve) => {
            const img = new Image();
            img.onload = () => {
                URL.revokeObjectURL(img.src);
                resolve({ width: img.width, height: img.height });
            };
            img.onerror = () => {
                URL.revokeObjectURL(img.src);
                resolve({ width: 0, height: 0 });
            };
            img.src = URL.createObjectURL(file);
        });
    }
}

var web = /*#__PURE__*/Object.freeze({
    __proto__: null,
    CapgoFilePickerWeb: CapgoFilePickerWeb
});

exports.CapgoFilePicker = CapgoFilePicker;
